### ISSUE TITLE
Explore sources of target synonyms

### ISSUE THREAD

**Issue by d0choa** [opentargets/platform#1393]
Getting a rich dictionary of gene/protein/target synonyms is becoming very relevant for the ePMC grounding among other things. As part of the target step refactoring we can consider including other sources for protein/gene synonyms.

One of the best aggregators of synonyms is gene cards. They mention what's the source of each synonym. [Example](https://www.genecards.org/cgi-bin/carddisp.pl?gene=AKT1)

Can we explore what would be the best sources to add and what's the cost/benefit of including them?

**Comment by DSuveges** [comment]
To generate gene objects in the platform ETL, we need to have a good collection of gene name synonyms and aliases. This dataset is also going to be used for grounding the ePMC ML result entities. 

**Sources to be considered:**

* HGNC complete gene set. [link](ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/json/hgnc_complete_set.json)
* NCBI gene info [link](ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz)

**Process:**

* Read both dataset, convert them to pandas dataframes.
* Re-shape tables where 3 columns will be preserved: `Ensembl` with Ensembl gene ids, `alias` containing any available alternative name/identifier/accession a gene was mention in the dataset and `alias_type` describing what type of alias it was eg. `hgnc_id`, `synonym`, `symbol` etc.
* As not all NCBI genes are mapped to ensembl, those genes, where only `hgnc` identifier is available is then mapped to Ensembl identifier using the hgnc dataset. 
* Genes in the hgnc dataset without ensembl mappings are discarded.
* Genes in the ncbi dataset, which could not be mapped to ensembl gene ids are discarded
* Concatenate the two datasets and remove duplicates.
* Identifiers and accessions can also be removed.

**Results:**

* In the HGNC dataset, of 39k unique genes there are 435k synonyms/cross references of which 174k synonyms.
* In the NCBI dataset there are 44k unique genes of which 35k could be mapped to Ensembl. There are 430k links, of which 290k alternative names/synonyms. 
* 6731 of the genes without Ensembl mapping in the NCBI dataset could be rescued with 49k synonyms/cross refs.
* There are still 2.5k genes that could not be mapped with 20k synonyms/cross refs.
* In the merged dataset there are synonyms/cross refs for 42k Ensembl genes. 
* The total number of unique links: 750k, the number of unique aliases: 257k.


**Alias types:** The list below shows all types of alternative names/synonyms/labels/accessions/identifiers:

Identifiers/accessions:

* `vega_id`
* `refseq_accession`
* `hgnc_id`
* `rgd_id`
* `omim_id`
* `mgd_id`
* `ucsc_id`
* `uniprot_ids`
* `ccds_id`
* `ena`
* `entrez_id`
* `imgt/gene-db_id`
* `mirbase_id`
* `ensembl_id`

Synonyms:

* `symbol`
* `name`
* `alias_symbol`
* `prev_name`
* `prev_symbol`
* `alias_name`
* `description`
* `symbol_from_nomenclature_authority`
* `full_name_from_nomenclature_authority`
* `other_designations`


**Head looks like this:**

```
Ensembl          alias                   alias_type
ENSG00000121410  OTTHUMG00000183507      vega_id
ENSG00000121410  NM_130786               refseq_accession
ENSG00000121410  5                       hgnc_id
ENSG00000121410  69417                   rgd_id
ENSG00000121410  138670                  omim_id
ENSG00000121410  A1BG                    symbol
ENSG00000121410  alpha-1-B glycoprotein  name
ENSG00000121410  2152878                 mgd_id
ENSG00000121410  uc002qsd.5              ucsc_id
```

Notebook is here: [link](https://github.com/DSuveges/random_notebooks/blob/master/issue-1393_getting_gene_name_synonyms/synonyms_from_HGNC.ipynb)
Dataset is here: [link](https://github.com/DSuveges/random_notebooks/blob/master/issue-1393_getting_gene_name_synonyms/target_mapping.tsv.gz)

**Comment by hammer** [comment]
You've likely already considered it but https://mygene.info is another useful resource in this area, and it's open source at https://github.com/biothings/mygene.info. cc @dhimmel who has worked with BioThings APIs quite a bit.

**Comment by dhimmel** [comment]
Thanks @hammer for looping me in. So my understanding of the goal here is to create a mapping between ensembl gene IDs and human-readable forms of the corresponding gene/protein/target. The solution by @DSuveges uses ncbigene to get an ensembl to gene symbol (including alternative symbols and names) mapping, but doesn't include protein/target names?

Are we sure this information is not already in ensembl? This is the query we developed to extract something link the NCBI `dbXrefs` from ensembl directly:

```sql
SELECT
  gene.stable_id AS ensembl_gene_id,
  external_db.db_name AS xref_source,
  xref.dbprimary_acc AS xref_accession,
  xref.display_label AS xref_label,
  xref.description AS xref_description,
  xref.info_type AS xref_info_type,
  object_xref.linkage_annotation AS xref_linkage_annotation
FROM gene
INNER JOIN object_xref
  ON gene.gene_id = object_xref.ensembl_id
  AND object_xref.ensembl_object_type = "Gene"
INNER JOIN xref
  ON xref.xref_id = object_xref.xref_id
INNER JOIN external_db
  ON xref.external_db_id = external_db.external_db_id
WHERE
  -- all genes were current when query was written, ensure this is always the case
  gene.is_current AND
  gene.biotype != "LRG_gene"
ORDER BY ensembl_gene_id, xref_source, xref_accession
```

Here's the number or records per `xref_source` broken down by `xref_info_type` on ensembl release 102 for human genes:

xref_source | DEPENDENT | DIRECT | MISC | All
-- | -- | -- | -- | --
ArrayExpress | 0 | 67139 | 0 | 67139
Clone_based_ensembl_gene | 0 | 0 | 24258 | 24258
DBASS3 | 0 | 188 | 0 | 188
DBASS5 | 0 | 277 | 0 | 277
ENS_LRG_gene | 0 | 862 | 0 | 862
EntrezGene | 28555 | 0 | 0 | 28555
HGNC | 3007 | 40412 | 0 | 43419
MIM_GENE | 17797 | 0 | 0 | 17797
MIM_MORBID | 7170 | 0 | 0 | 7170
RFAM | 0 | 4969 | 0 | 4969
Reactome_gene | 0 | 132036 | 0 | 132036
Uniprot_gn | 73181 | 0 | 0 | 73181
WikiGene | 28555 | 0 | 0 | 28555
miRBase | 0 | 0 | 2115 | 2115
All | 158265 | 245883 | 26373 | 430521

So it looks perhaps like there is no record of alternative symbols? However looking on the Ensembl website for [ENSG00000142208](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000142208;r=14:104769349-104795751) (AKT1), there are several synonyms listed: "AKT, PKB, PRKBA, RAC, RAC-alpha". But it's still less comprehensive than gene cards. (Too bad gene cards [doesn't](https://www.genecards.org/Guide/Datasets) make their database dumps readily available).

More comprehensive cataloging of alternative gene symbols is not always helpful, especially if there is no way to know how prevalent the usage of a given alternative symbol is. I'd be curious to know when you subset your ncbigene-derived xref table for just symbols (official and aliases), what percent of symbols have a collision (do not unambiguously map to a single ensembl representative gene)? Answering this question is complicated by the difficultly of determining whether an ensembl gene is representative or is an alternative sequence that should be collapsed with another ensembl gene. If the goal is text mining, the more symbol xrefs might lead to more ambiguous mappings and actually reduce your overall ability to tag gene entities in text.

Side note regarding the following line, I've been [burned in the past](https://think-lab.github.io/d/34/#9) by not filtering `Homo_sapiens.gene_info.gz` for `tax_id == 9606`:

```py
ncbi_df = pd.read_csv('Homo_sapiens.gene_info.gz', compression='infer', sep='\t')
```

Looks like you are filtering `gene2ensembl.gz` for 9606, so you might already be excluding Neanderthal and Denisovan genes.

Noting [cognoma/genes](https://github.com/cognoma/genes/blob/ec0e5e5fc0066a6c1829866cccab882b7888d709/2.process.py#L77-L112), where we processed `Homo_sapiens.gene_info.gz` for xrefs, in case you want another implementation to compare to.

> You've likely already considered it but https://mygene.info is another useful resource in this area

I'm not sure how much benefit there is to using mygene here as opposed to directly ingesting the ncbigene dump. One issue with MyGene is that gene records are indexed by their NCBI gene if that exists, but otherwise their ensembl ID. So MyGene is a fusion of ncbi and ensembl in a way that can make it difficult to consistently operate on these resources separately. Furthermore, there are some issues with inconsistent return types of plural fields.

This being said, if MyGene is able to provide quick access to the protein names for a given gene, that seems like a major benefit.

Apologies for the lengthy reply. I'm interested to follow the progress here and hopefully this is of some help.

**Comment by d0choa** [comment]
Thanks both! 

As @dhimmel mentioned, the Xrefs are already quite good in Ensembl for the vast majority of our use-cases, so we might stick with that.

The original intention was to increase the synonym coverage. We are currently doing the grounding of NER recognised genes/proteins, so a rich dictionary of terms is important to get a decent coverage. The list we can retrieve from Ensembl or Uniprot does seem limited and we would benefit from adding other sources. By looking at a few random GeneCards entries, NCBI Entrez looks like the most promising source of synonyms.

However, @dhimmel comment on clashes is also true and something I have discussed with @DSuveges as well. We might need to impose some sort of precedence similar to the EFO `hasExactSynonym`, `hasCloseSynonym`, etc. We will report some numbers on clashes as soon as we have a first pass on this. @JarrodBaker is helping with this.

**Comment by DSuveges** [comment]
@hammer, thanks for comment! I was considering mygene.info, but an API solution wouldn't be really good for us, and when I didn't find an easily downloadable set, I abandoned that resource, without digging deeper where they sourcing their data.

As @dhimmel said, I didn't want to be overly complete to avoid too many ambiguous mappings (there are already 4k+ labels mapped to more than one Ensembl IDs.). Therefore I decided to go for the two probably most notable sources: HGNC and NCBI.

@dhimmel Thanks for the comments! You were right, I was using the already filtered NCBI gene info set, containing only Homo sapiens genes. We'll consider how to handle Neanderthal and Denisovan genes. We'll also take a look at how we could include more protein centered labels too. Extracting data directly from Ensembl db also a great suggestion (the query you shared looks amazing). For the next iteration I'll certainly check it. Considering how many other sources Ensembl integrates, we can expect a major increase in labels.

I also noticed the ambiguity in the mappings across identifiers: more than one Ensembl to one HGNC: 139;  more than one Ensembl to one NCBI: 59 genes. We'll need to look into consequences of this ambiguity and the scale how frequently they used.

Thank you guys for the great suggestions! We are in the beginning of this journey, any input welcome.
